### ISSUE TITLE
Confusing pagination result with keyArgs in FieldPolicy

### ISSUE THREAD

**Issue by mschipperheyn** [apollographql/apollo-client#6537]
I have a pagination FieldPolicy like below:

```
const offsetLimitPagination = (
	keyArgs = ['filter', 'sort'],
	opts = {
		concat: true,
	},
) => {
	return {
		keyArgs,
		merge(existing = makeEmptyData(), incoming, { args }) {
			let prefix = existing.edges.slice(0);
			let suffix = [];
			const incomingEdges = incoming.edges.slice(0);

			const pageInfo = {
				...existing.pageInfo,
				...incoming.pageInfo,
			};
			let edges;
			if (opts.concat) {
				edges = [...prefix, ...incomingEdges, ...suffix];
			} else {
				edges = [...incomingEdges];
			}
			const data = {
				...existing,
				...incoming,
				edges,
				pageInfo,
			};
			return data;
		},
	};
};

function makeEmptyData() {
	return {
		edges: [],
		pageInfo: {
			hasPreviousPage: false,
			hasNextPage: true,
			page: 0,
			pageSize: 100,
		},
		totalCount: 0,
	};
}
```

Type policy
```
admin__social__groups: offsetLimitPagination(undefined, { concat: false })
```
I have a query that looks like this:

```
query Admin__Social__Groups(
		$page: Int = 0
		$pageSize: Int = 10
		$filter: String
		$sort: String
		$isCursor: Boolean = false
		$before: String
		$after: String
	) {
		admin__social__groups(
			pageSize: $pageSize
			filter: $filter
			sort: $sort
			page: $page
			isCursor: $isCursor
			before: $before
			after: $after
		) {
			edges {
				node {
					...groupFragment
					status {
						hide
						review
						promoted
						autoJoin
					}
					updatedAt
				}
				cursor
			}
			pageInfo {
				...basePageInfoFragment
			}
			totalCount
		}
	}
```
I call `fetchMore({ variables: { page: 1 } })`

When I use default keyArgs `['filter', 'sort']`, the pagination works as expected. But when I use keyArgs
```
[
	'page',
	'before',
	'after',
	'filter',
	'sort',
	'pageSize',
	'isCursor',
];
```
which feels more correct, bc all keys influence outcome, pagination stops working. The data is retrieved and the merge function does get processed correctly, but the screen doesn't update. No rerendering occurs.

**Intended outcome:**
If something is returned from the merge function, it should be rendered. 

**Actual outcome:**
Query doesn't rerender. The old result remains isn't replaced. My guess is that the cache key processing gets confused because of the undefined values? The array that leads to failure does include a changed key (`page`). Whereas the default only contains undefined keys or `['sort','filter']` (unused).

My guess is that the cache key is not calculated correctly. Or that I misunderstand how this is supposed to work.
Whenever any of the parameters passed as keyArgs actually contains a value in the query, merge completes without rerendering and essentially fails. 

**How to reproduce the issue:**
See above

**Versions**
rc.10

**Comment by benjamn** [comment]
I think you have a misunderstanding of what `keyArgs` are for. You do not need to include args in `keyArgs` just because they influence the outcome. You only need to include arguments that identify completely separate instances of the field data, such as a search query whose results have nothing to do with the results of other search queries. By including args like `sort` and `filter` in `keyArgs`, you're missing out on the opportunity to maintain an unsorted, unfiltered, authoritative list, which is then sorted or filtered (or whatever) by the `read` function. Including more key arguments just means your data will be more fragmented, and you will have less opportunity to benefit from a single combined copy of the data. From a glance, none of the arguments you mentioned belong in `keyArgs`—again, not because they aren't important to the outcome, but because `keyArgs` can only use the arguments to separate the data, whereas a `read` function can actually interpret the arguments and return different views of the same underlying data. In principle, `read` and `merge` functions can take full responsibility for interpretation of arguments, so you never really need `keyArgs` (that is, you can always get away with using `keyArgs: false`). It's just useful sometimes to move certain arguments into `keyArgs` so that `read` and `merge` don't have to worry about them.

**Comment by mschipperheyn** [comment]
@benjamn It turns out we find using the typePolicy for pagination extremely tricky. Some scenarios we've run into:

- defining typeArgs as undefined can lead to unexpected results when you are using the same query with varying page sizes. If we don't add `pageSize` as a keyArg, a cached list of 12 items will show up on a location where pageSize = 8.
- Misunderstanding keyArgs as I did above and adding the wrong keyArgs that have values will lead to pagination not working at all by not displaying the result of a merged list as described above (still not clear why). 

All in all, you can't seem to always be able to get away with `keyArgs: false`

**Comment by benjamn** [comment]
I did not mean that `keyArgs: false` always just works without additional effort in the `read` and `merge` functions to handle the arguments in appropriate ways. However, I stand by my advice that none of the arguments you mentioned above should be part of the `keyArgs` field key.

Here's a comment from another issue that might be worth reading and thinking about: https://github.com/apollographql/apollo-client/issues/6481#issuecomment-648890620

I honestly think you need to be patient, keep interrogating your assumptions, and revisit my explanations with fresh eyes, before assuming that the code you've written should be working. We made the decision to provide helper functions like `relayStylePagination` because we realized how tricky these field policies can be. They're tricky in part because they're so flexible, and every little decision matters, but also because Relay-style pagination is just hideously over-engineered and unnecessarily complicated, and any attempt to abstract away those complexities is going to be sensitive to the many edge cases they present.

As always, I'll be here to answer questions.

**Comment by mschipperheyn** [comment]
Fair enough, I'll keep trying to wrap my head around it. I'm also seeing some refetchQueries not updating the UI that I'm now wondering if they might be related. Thanks for your patience.

**Comment by mschipperheyn** [comment]
So, if I understand all of this correctly, implementing a `read` function basically presumes you want to manipulate data from the cache in an intelligent way. Blindly copying that got me into trouble here then. I should just not have a read function and just use a merge function.

I'm also noticing as I remove the `read` function that the use of keyArgs doesn't seem to matter anymore. With the `read` function I was having issues with two different queries showing the same results if I didn't add e.g. a `userId` as a `keyArg`. 

I still think this is odd and prob shouldn't happen or perhaps should show some kind of warning "Two identical cache results have different cache keys. Check your type policy" or something along those lines. But for now, I'm good. Thanks.
